### ISSUE TITLE
Critical vectorization issue

### ISSUE THREAD

**Issue by Bonnevie** [Bonnevie/rebar#2]
Assuming the `loss` and `logp` tensors passed to `RELAX` are meaned over the number of samples leads to problems when
```
approx_gap = (loss - weight*conditional_control)
[...]
score_grad = approx * score
```
is encountered, as the product in `score_grad` (meant to be element-wise over samples) becomes a product of two means, expanding into a quadratic number of terms, inducing significant bias that ironically gets more severe with the number of samples used.

**Comment by Bonnevie** [comment]
This has been fixed today using the graph editor. By calculating
```
        approx_gap = (loss - weight*conditional_control)
        blocked_approx_gap = tf.stop_gradient(approx_gap)
        with tf.name_scope("score"):
            scores = tf.gradients(tf.reduce_mean(logp*blocked_approx_gap), hard_params)
        scores = tf.contrib.graph_editor.graph_replace(scores, {blocked_approx_gap: approx_gap})
```
the element-wise mean is thus applied prior to applying the gradient to the log distribution, and the combination of `stop_gradient` and `graph_replace` allows us to prevent the gradient from applying itself to `approx_gap`, before substituting the function with the proper derivative properties back in again (important for variance gradient).
